### DISCUSSION TITLE
pag-flutter插件兼容鸿蒙

### DISCUSSION THREAD

**Discussion by zuojiganga** [Tencent/libpag#2526] [Q&A] [answered]
pag-flutter兼容鸿蒙，参照安卓、iOS逻辑写，画面黑屏不出来，使用texturesRegistry加载png格式图片是可以的，texturesRegistry逻辑应该是对的，pag文件不行，如下：
let textureId = this.texturesRegistry?.getTextureId()
      let surfaceId = this.texturesRegistry?.registerTexture(textureId).getSurfaceId()
      let manager = getContext(this).resourceManager;
      let pagFile = PAGFile.LoadFromAssets(manager, pagPath);
      this.pagFile = pagFile
      this.player =  new PAGPlayer()
      this.player?.setComposition(this.pagFile)
      // this.surface = PAGSurface.MakeOffscreen(this.pagFile?.width(), this.pagFile?.height())
      let surface = PAGSurface.FromSurfaceID(surfaceId)
      this.surface = surface
      this.player?.setSurface(this.surface)
      this.player?.setProgress(this.initProgress)
      this.player?.flush()
请问，哪里代码逻辑有什么问题？

**Comment by domchen** [comment]
检查一下surface创建的返回值是不是空，或者有没有打印失败信息。如果创建失败了，去检查surfaceID你分配内存了没有。就是从上面要能用鸿蒙接口查询出有效宽高才行。PAG不帮你分配内存。

**Comment by zuojiganga** [comment]
![WechatIMG423](https://github.com/user-attachments/assets/b984aa8b-cbe1-4820-b245-23e9966789cf)
打印了，如上传图片所示，PAGFile、PAGPlayer、PAGSurface都是有值的。PAGFile的宽高也有。再问下，假如以上代码逻辑无问题，是不是就可以出来画面？还是说除了以上代码，还需要继续写动画animateTo逻辑？

**Reply by domchen** [reply]
要自己接源码模式调试一下了，看看挂在哪个步骤。

**Reply by act262** [reply]
@zuojiganga  1. 需要实现animator 驱动刷新；2. 鸿蒙flutter下的Texture对应的window窗口大小为0导致没有效果，需要手动设置OHNativeWindow宽高; 3. 鸿蒙flutter不支持 传递double类型报错，      

```
 let callback = new HashMap<string, Object>();
      callback.set(FlutterPagPlugin._argumentTextureId, textureId);
      // TODO: ohos not support `double` type, so not set width and height here
      // https://gitee.com/openharmony-sig/flutter_flutter/issues/I9QR10?from=project-issue
      // callback.set(FlutterPagPlugin._argumentWidth, width);
      // callback.set(FlutterPagPlugin._argumentHeight, height);
```

**Reply by zuojiganga** [reply]
1. 需要实现animator 驱动刷新:
animator如何关联要执行的组件呢？此处是一个PAGPlayer。 2和3点的相关逻辑是有的。

**Reply by zuojiganga** [reply]
2. 鸿蒙flutter下的Texture对应的window窗口大小为0导致没有效果，需要手动设置OHNativeWindow宽高;
问：是如下鸿蒙返回flutter纹理的宽高吗？
result.success({
      "textureId":textureId,
      "width":width,
      "height":height
    }

**Reply by act262** [reply, marked as answer]
可以参考下这个实现 https://github.com/libpag/pag-flutter/pull/51

**Reply by zuojiganga** [reply]
https://github.com/libpag/pag-flutter/pull/51  example运行本地pag文件可以加载出来，网络pag文件加载不出来

**Reply by act262** [reply]
已经可以了，修复了传递double类型转换出错问题 （force push），你重新更新下代码

**Reply by zuojiganga** [reply]
您好，我是模块化工程，在模块中使用https://github.com/libpag/pag-flutter/pull/51插件，传了package参数，pagfile初始化不成功，pag文件路径好像不对

**Reply by act262** [reply]
模块化工程是指原生鸿蒙模块还是flutter模块？麻烦提供个sample示例呢

**Comment by zuojiganga** [comment]
<img width="358" alt="image" src="https://github.com/user-attachments/assets/42799a55-a102-4ea9-9dd7-d18f4e49dd2e">

flutter端模块，如上图：目录中pag-flutter、res、third_plugin是同级的三个模块，xxx.pag文件在res模块中资源目录下，third_plugin模块某页面中调用了pag-flutter模块中 PAGView.asset()。

**Comment by zuojiganga** [comment]
<img width="822" alt="image" src="https://github.com/user-attachments/assets/c51b07a9-a832-4719-b7d7-13f2415f575d">
已经修复了，此处添加代码 if (flutterPackage != undefined && (flutterPackage?.length > 0)) {
          assetName = 'packages/' + flutterPackage  + '/'+ assetName 
      }

**Reply by act262** [reply]
package 下资源无效问题已更新 flutter engine ， pag-flutter 无需更新。

See this https://gitee.com/openharmony-sig/flutter_engine/pulls/523

**Comment by zuojiganga** [comment]
有新的问题，app的一级页面（底部tab对应页面）有一个pag动画在播放，当我跳转到二级页面，再返回，然后每一秒输出一条：FlutterNapi --> Tried to send a platform message to Flutter, but FlutterNapi was detached from native C++. Could not send. Channel: flutter_pag_plugin. Response ID: 88。如上多操作几次，然后每一秒输出的越多。最重要的是页面开始卡顿，甚至闪退。

**Reply by act262** [reply]
- 你需要看下有没有调用到release 方法，在PagView.dispose 方法中被调用。

- 内存上涨卡顿问题目前来看是flutter engine 层导致的，需要时间验证。

**Comment by zuojiganga** [comment]
通过输出日志查看，在PagView.dispose 方法中的release 方法是调用了，不是这个原因。新发现一点希望有帮助，一级页面跳转二级页面，使用路由跳转 router.pushUrl()没问题，改用Navigation() {}组件跳转就有问题（动画几乎不动），必现。

**Comment by zuojiganga** [comment]
<img width="1354" alt="image" src="https://github.com/user-attachments/assets/34439fdc-5244-4b22-9448-cd418f5c2c14">
更正一点，router.pushUrl() 和Navigation() {}  跳转到二级页面入栈都是没有问题，再跳转第三级、第四级....都是没问题的，但是，只要pop返回上级页面，就会卡着不动,然后闪退，报错如上图。使用场景是：每级页面都是使用鸿蒙原生导航栏，主页面内容是flutter组件（使用了pagView）。使用flutterEngineGroup路由管理，flutterEgine不释放，pagView也不释放
